sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -280,6 +283,7 @@ def get(self, instruction_id, bundle_descriptor_id):
     try:
       # pop() is threadsafe
       processor = self.cached_bundle_processors[bundle_descriptor_id].pop()
+      self.last_access_time[bundle_descriptor_id] = time.time()
     except IndexError:
```

Comment:
This won't update the access time when we first create the processor in the except block.

Reply:
This is expected. The `last_access_time` represents the last time the `cached_bundle_processors` for some `bundle_descriptor` is accessed. If it exceeds the time limit, the remaining bundle processors cached in the `cached_bundle_processors` will be shutdown. What's your thought?

Reply:
In my eyes, the first time it is accessed is when it is created.

Reply:
The strategy here is that the bundle processors which are unused after an amount of time will be shutdown. When a bundle processor is created in the exception block, there are no cached(unused) bundle processors. The bundle processors become unused only when they are added to the cached bundle processors list. What about rename last_access_time to `cached_bundle_processors_last_access_time` to make it more clear?

Reply:
I still do not understand, the method is named `get`, so we access the bundle processor independently of whether we create it or not. It is cached, regardless of whether it is created and added to the cache, or retrieved from the cache. 

Logically, you might want to update the time when putting the processor into the cache. That would be in `release`.

What is the advantage of updating the time here? It should be sufficient to update it in `release`, directly before putting it back.

Reply:
Hmm, If a bundle processor is retrieved from the cache, there is a high possibility that the remaining cached bundle processors will be needed in the future and so the last access time is updated.

If the bundle processor is newly created, it means that the cached bundle processor list is empty. This is the main reason that the last access time is only updated when bundle processor is retrieved from the cache. However, I think that it does no harm to update the last access time in both cases if it makes the code more readable.

What do you think?

Reply:
>If the bundle processor is newly created, it means that the cached bundle processor list is empty. This is the main reason that the last access time is only updated when bundle processor is retrieved from the cache. 

Consider the case where we just have a single bundle processor. When we call get for the first time, we won't update the last-used time. However, every time we retrieve it afterwards, we will update the time, but the list of cached bundle processors will remain empty.

I think we should either (1) always update the last-used timestamp in `get`, regardless of creation or (2) update it only on `release`.

I'm leaning towards (2) because while a bundle processor is in-use, it can't be removed anyways. We update the timestamp when we put it back in `release`.

Reply:
Regarding the single bundle processor case, it doesn't harm to update the time as the cached bundle processors is empty. However, in cases where there are multiple bundle processors, it will update the time for the remaining cached bundle processors and so improve the cache hit rate. I think this is main difference between solution (1) and (2). However, I'm fine with both solutions as I think both of them work. Will update the PR according to solution (2) if you are favor of it according to your experience. What's your thought?